analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/ObjectsShouldNotBeDisposedMoreThanOnceBase.cs

```diff
@@ -0,0 +1,49 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.SymbolicExecution.Constraints;
+
+namespace SonarAnalyzer.SymbolicExecution.Roslyn.RuleChecks;
+
+public abstract class ObjectsShouldNotBeDisposedMoreThanOnceBase : SymbolicRuleCheck
+{
+    protected const string DiagnosticId = "S3966";
+    protected const string MessageFormat = "Refactor this code to make sure '{0}' is disposed only once.";
+
+    private static readonly string[] DisposeMethods = { "Dispose", "DisposeAsync", "Close" };
```

Comment:
We're missing UTs for `DisposeAsync`, `Close`, and their combinations with `Dispose`

Reply:
I did not add a UT for "close". I removed it.

I went a bit deeper now and I saw that `close` implementations are not expected to have consistent behavior. For example, there are [different behaviors on streams](https://learn.microsoft.com/en-us/dotnet/api/system.io.stream.close?view=net-7.0) and [different on DB Connections](https://learn.microsoft.com/en-us/dotnet/api/system.data.sqlclient.sqlconnection.close?view=dotnet-plat-ext-7.0).
And maybe there are other cases as well.

If I just add close I'll probably raise false positives (for example closing twice on a connection that has been re-opened is not bad).

I was thinking to open an issue for supporting the relevant close cases only - because it needs research and removing it for now.

WDYT?